File: 0948-sort-an-array/0948-sort-an-array.java
class Solution {

    public int partition(int[] arr, int low, int high) {
        // Use the middle element as the pivot to avoid worst-case performance
        int mid = low + (high - low) / 2;
        int pivot = arr[mid];
        
        // Move pivot to the end
        swap(arr, mid, high);
        
        int i = low - 1;
        
        for (int j = low; j < high; j++) {
            if (arr[j] < pivot) {
                i++;
                swap(arr, i, j);
            }
        }
        i++;
        swap(arr, i, high);
        return i;
    }

    public void quickSort(int[] arr, int low, int high) {
        while (low < high) {
            int pidx = partition(arr, low, high);
            
            // Sort the smaller partition first to optimize recursion depth
            if (pidx - low < high - pidx) {
                quickSort(arr, low, pidx - 1);
                low = pidx + 1;  // Tail recursion for the larger partition
            } else {
                quickSort(arr, pidx + 1, high);
                high = pidx - 1;  // Tail recursion for the smaller partition
            }
        }
    }

    // Helper function to swap elements
    private void swap(int[] arr, int i, int j) {
        int temp = arr[i];
        arr[i] = arr[j];
        arr[j] = temp;
    }

    public int[] sortArray(int[] nums) {
        // quickSort(nums, 0, nums.length - 1);
        Arrays.sort(nums);
        return nums;
    }
}

File: 0948-sort-an-array/0948-sort-an-array.py
# class Solution(object):
#     def partition(self, nums, low, high):
#         pivot = nums[high]  # Set pivot to be the last element
#         i = low - 1

#         for j in range(low, high):
#             if nums[j] < pivot:
#                 i += 1
#                 nums[i], nums[j] = nums[j], nums[i]

#         # Place pivot element in the correct position
#         nums[i + 1], nums[high] = nums[high], nums[i + 1]
#         return i + 1

#     def quick(self, nums, low, high):
#         if low < high:
#             pidx = self.partition(nums, low, high)
#             self.quick(nums, low, pidx - 1)
#             self.quick(nums, pidx + 1, high)

#     def sortArray(self, nums):
#         """
#         :type nums: List[int]
#         :rtype: List[int]
#         """
#         self.quick(nums, 0, len(nums) - 1)
#         return nums

class Solution(object):
    def sortArray(self, nums):
        """
        :type nums: List[int]
        :rtype: List[int]
        """
        nums.sort()
        return nums